### PR TITLE
fix(developer): remove empty touch rows on save

### LINKED ISSUE #5327: bug(web): touch keyboard with empty row causes TypeError: Cannot read property 'length' of undefined
Sentry Issue: [KEYMAN-DEVELOPER-8K](https://sentry.keyman.com/organizations/keyman/issues/6781/?referrer=github_integration)

```
TypeError: Cannot read property 'length' of undefined
  at Function.ActiveLayer.polyfill (/resource/keymanweb.js:4733:50)
  at Function.ActiveLayout.polyfill (/resource/keymanweb.js:4973:37)
  at Keyboard.layout (/resource/keymanweb.js:6375:64)
  at OSKManager._Load (/resource/keymanweb.js:29261:62)
  at HTMLScriptElement.<anonymous> (/resource/keymanweb.js:23580:37)
```

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #5327.

# User Testing

This is in the touch layout editor.

* **TEST_BASELINE**: Run through basic editing scenarios for any touch layout, and make sure that files continue to load and save and be editable.
* **TEST_CLEANUP**: Verify that empty rows are removed.

  1. Using the afghan_turkmen.zip attached to #5327, look at afghan_turkmen_1_1.keyman-touch-layout **in a text editor**.
  2. In the text editor (not in Keyman Developer), verify that the `phone` layout has a blank first row:

      ```json
      {
        "id": 1,
        "key": []
      },
      ```
  3. Load afghan_turkmen_1_1.kmn, make a single change to the touch layout, and save.
  4. In the text editor, reload the file, and verify that the `phone` layout in afghan_turkmen.keyman-touch-layout no longer has a blank first row.

### CODE REVIEW

File: windows/src/developer/TIKE/xml/layoutbuilder/builder.js [hunk 1]
```diff
@@ -2263,6 +2268,23 @@ $(function () {
     );
   };
 
+  builder.removeEmptyRows = function() {
+    var json = JSON.stringify(KVKL, null, '  ');
+    for (var platform in KVKL) {
+      for (let layer of KVKL[platform].layer) {
+        let newRow = [], n = 1;
+        for(let row of layer.row) {
```
Comment (darcywong00) [inline windows/src/developer/TIKE/xml/layoutbuilder/builder.js:2276]: if `layer.row` is `undefined`, will l.2283 mean `layer.row = []` which gets saved?

Comment (mcdurdin) [reply windows/src/developer/TIKE/xml/layoutbuilder/builder.js:2276]: `layer.row` should never be `undefined` though!

**Review by darcywong00 — APPROVED** [review]
lgtm

File: windows/src/developer/TIKE/xml/layoutbuilder/builder.js (hunk 1 above)
Comment (jahorton) [reply windows/src/developer/TIKE/xml/layoutbuilder/builder.js:2276]: > should

I agree... but "should" and "will" are unfortunately words with very different meanings.

Comment (mcdurdin) [reply windows/src/developer/TIKE/xml/layoutbuilder/builder.js:2276]: If layer.row was undefined, then we'd be seeing errors all over the place.

The touch layout editor is another piece of code that has grown past its original remit and needs refactoring. A job for another day.

**Comment by MakaraSok** [comment]
- **TEST_BASELINE: PASSED** The touch layout edition saved and continued to work.
- **TEST_CLEANUP: BLOCKED** The touch layout file does have the code block when open in `Notepad++` .
   

   Once loaded the .kmn, the blank row went away when opened in Keyman Developer Touch Layout Code view. Nothing has been done to the touch layout, not a slight edit in there.

   

   Here is how the phone layout look without any edition:

    

   No blank row visible in the Design view.

    Since the touch layout looks OK in both the Design and Code views and I was not able to carry out step `#iii`, this is marked as BLOCKED for now.

**Comment by mcdurdin** [comment]
> * Once loaded the .kmn, the blank row went away when opened in Keyman Developer Touch Layout Code view. Nothing has been done to the touch layout, not a slight edit in there.

OK, updated the steps slightly :grin: -- check the content of the file in a separate text editor -- Keyman Developer does the cleanup on load but won't save changes until you tell it to.

**Comment by MakaraSok** [comment]
* **TEST_CLEANUP: PASSED** The empty first row on touch layout does disappear after making a change to the layout in KD.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.121-alpha](https://keyman.com/downloads/releases/alpha/15.0.121)

**Pull request merged**